// wppconnect/dist/api/model/enum/message-type.js
"use strict";
/*
 * This file is part of WPPConnect.
 *
 * WPPConnect is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * WPPConnect is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with WPPConnect.  If not, see <https://www.gnu.org/licenses/>.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaType = exports.MessageType = void 0;
var MessageType;
(function (MessageType) {
    MessageType["TEXT"] = "chat";
    MessageType["AUDIO"] = "audio";
    MessageType["VOICE"] = "ptt";
    MessageType["IMAGE"] = "image";
    MessageType["VIDEO"] = "video";
    MessageType["DOCUMENT"] = "document";
    MessageType["STICKER"] = "sticker";
    MessageType["LOCATION"] = "location";
    MessageType["CONTACT_CARD"] = "vcard";
    MessageType["CONTACT_CARD_MULTI"] = "multi_vcard";
    MessageType["REVOKED"] = "revoked";
    MessageType["UNKNOWN"] = "unknown";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
var MediaType;
(function (MediaType) {
    MediaType["IMAGE"] = "Image";
    MediaType["VIDEO"] = "Video";
    MediaType["AUDIO"] = "Audio";
    MediaType["PTT"] = "Audio";
    MediaType["DOCUMENT"] = "Document";
    MediaType["STICKER"] = "Image";
})(MediaType = exports.MediaType || (exports.MediaType = {}));
