// wppconnect/dist/api/model/enum/socket-state.js
"use strict";
/*
 * This file is part of WPPConnect.
 *
 * WPPConnect is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * WPPConnect is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with WPPConnect.  If not, see <https://www.gnu.org/licenses/>.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocketStream = exports.SocketState = void 0;
var SocketState;
(function (SocketState) {
    SocketState["CONFLICT"] = "CONFLICT";
    SocketState["CONNECTED"] = "CONNECTED";
    SocketState["DEPRECATED_VERSION"] = "DEPRECATED_VERSION";
    SocketState["OPENING"] = "OPENING";
    SocketState["PAIRING"] = "PAIRING";
    SocketState["PROXYBLOCK"] = "PROXYBLOCK";
    SocketState["SMB_TOS_BLOCK"] = "SMB_TOS_BLOCK";
    SocketState["TIMEOUT"] = "TIMEOUT";
    SocketState["TOS_BLOCK"] = "TOS_BLOCK";
    SocketState["UNLAUNCHED"] = "UNLAUNCHED";
    SocketState["UNPAIRED"] = "UNPAIRED";
    SocketState["UNPAIRED_IDLE"] = "UNPAIRED_IDLE";
})(SocketState = exports.SocketState || (exports.SocketState = {}));
var SocketStream;
(function (SocketStream) {
    SocketStream["CONNECTED"] = "CONNECTED";
    SocketStream["DISCONNECTED"] = "DISCONNECTED";
    SocketStream["RESUMING"] = "RESUMING";
    SocketStream["SYNCING"] = "SYNCING";
})(SocketStream = exports.SocketStream || (exports.SocketStream = {}));
