// wppconnect/dist/controllers/browser.js
"use strict";
/*
 * This file is part of WPPConnect.
 *
 * WPPConnect is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * WPPConnect is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with WPPConnect.  If not, see <https://www.gnu.org/licenses/>.
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initBrowser = exports.injectApi = exports.initWhatsapp = void 0;
var ChromeLauncher = require("chrome-launcher");
var path = require("path");
var axios_1 = require("axios");
var puppeteer_extra_1 = require("puppeteer-extra");
var puppeteer_config_1 = require("../config/puppeteer.config");
var StealthPlugin = require("puppeteer-extra-plugin-stealth");
var auth_1 = require("./auth");
var WAuserAgente_1 = require("../config/WAuserAgente");
var websocket_1 = require("./websocket");
function initWhatsapp(session, options, browser, token) {
    return __awaiter(this, void 0, void 0, function () {
        var waPage, timeout;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getWhatsappPage(browser)];
                case 1:
                    waPage = _a.sent();
                    if (!(waPage != null)) return [3 /*break*/, 5];
                    // Auth with token
                    return [4 /*yield*/, auth_1.auth_InjectToken(waPage, session, options, token)];
                case 2:
                    // Auth with token
                    _a.sent();
                    return [4 /*yield*/, waPage.setUserAgent(WAuserAgente_1.useragentOverride)];
                case 3:
                    _a.sent();
                    timeout = 2 * 1000;
                    return [4 /*yield*/, Promise.race([
                            waPage.goto(puppeteer_config_1.puppeteerConfig.whatsappUrl, { timeout: timeout }).catch(function () { }),
                            waPage.waitForSelector('body', { timeout: timeout }).catch(function () { }),
                        ])];
                case 4:
                    _a.sent();
                    return [2 /*return*/, waPage];
                case 5: return [2 /*return*/, false];
            }
        });
    });
}
exports.initWhatsapp = initWhatsapp;
function injectApi(page) {
    return __awaiter(this, void 0, void 0, function () {
        var injected;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, page
                        .evaluate(function () {
                        // @ts-ignore
                        return (typeof window.WAPI !== 'undefined' &&
                            typeof window.Store !== 'undefined');
                    })
                        .catch(function () { return false; })];
                case 1:
                    injected = _a.sent();
                    if (injected) {
                        return [2 /*return*/];
                    }
                    return [4 /*yield*/, page.addScriptTag({
                            path: require.resolve(path.join(__dirname, '../lib/wapi', 'wapi.js')),
                        })];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, page.addScriptTag({
                            path: require.resolve(path.join(__dirname, '../lib/middleware', 'middleware.js')),
                        })];
                case 3:
                    _a.sent();
                    // Make sure WAPI is initialized
                    return [4 /*yield*/, page.waitForFunction(function () {
                            // @ts-ignore
                            return (typeof window.WAPI !== 'undefined' && typeof window.Store !== 'undefined');
                        })];
                case 4:
                    // Make sure WAPI is initialized
                    _a.sent();
                    return [2 /*return*/, page];
            }
        });
    });
}
exports.injectApi = injectApi;
/**
 * Initializes browser, will try to use chrome as default
 * @param session
 */
function initBrowser(session, options, extras) {
    if (extras === void 0) { extras = {}; }
    return __awaiter(this, void 0, void 0, function () {
        var chromePath, browser, transport;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (options.useChrome) {
                        chromePath = getChrome();
                        if (chromePath) {
                            extras = __assign(__assign({}, extras), { executablePath: chromePath });
                        }
                        else {
                            console.log('Chrome not found, using chromium');
                            extras = {};
                        }
                    }
                    // Use stealth plugin to avoid being detected as a bot
                    puppeteer_extra_1.default.use(StealthPlugin());
                    browser = null;
                    if (!(options.browserWS && options.browserWS != '')) return [3 /*break*/, 3];
                    return [4 /*yield*/, getTransport(options.browserWS)];
                case 1:
                    transport = _a.sent();
                    return [4 /*yield*/, puppeteer_extra_1.default.connect({ transport: transport })];
                case 2:
                    browser = _a.sent();
                    return [3 /*break*/, 5];
                case 3: return [4 /*yield*/, puppeteer_extra_1.default.launch(__assign(__assign({ headless: options.headless, devtools: options.devtools, args: options.browserArgs
                            ? options.browserArgs
                            : __spreadArrays(puppeteer_config_1.puppeteerConfig.chromiumArgs) }, options.puppeteerOptions), extras))];
                case 4:
                    browser = _a.sent();
                    _a.label = 5;
                case 5: return [2 /*return*/, browser];
            }
        });
    });
}
exports.initBrowser = initBrowser;
function getWhatsappPage(browser) {
    return __awaiter(this, void 0, void 0, function () {
        var pages;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    pages = null;
                    return [4 /*yield*/, browser
                            .pages()
                            .then(function (e) {
                            console.assert(e.length > 0);
                            pages = e[0];
                        })
                            .catch(function () { })];
                case 1:
                    _a.sent();
                    return [2 /*return*/, pages];
            }
        });
    });
}
/**
 * Retrieves chrome instance path
 */
function getChrome() {
    try {
        var chromeInstalations = ChromeLauncher.Launcher.getInstallations();
        return chromeInstalations[0];
    }
    catch (error) {
        return undefined;
    }
}
function getTransport(browserWS) {
    return __awaiter(this, void 0, void 0, function () {
        var error, e_1, endpointURL, data, e_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    error = null;
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, websocket_1.WebSocketTransport.create(browserWS, 10000)];
                case 2: return [2 /*return*/, _a.sent()];
                case 3:
                    e_1 = _a.sent();
                    error = e_1;
                    return [3 /*break*/, 4];
                case 4:
                    _a.trys.push([4, 7, , 8]);
                    endpointURL = browserWS.replace(/ws(s)?:/, 'http$1:') + '/json/version';
                    return [4 /*yield*/, axios_1.default.get(endpointURL).then(function (r) { return r.data; })];
                case 5:
                    data = _a.sent();
                    return [4 /*yield*/, websocket_1.WebSocketTransport.create(data.webSocketDebuggerUrl, 10000)];
                case 6: return [2 /*return*/, _a.sent()];
                case 7:
                    e_2 = _a.sent();
                    return [3 /*break*/, 8];
                case 8: 
                // Throw first error
                throw error;
            }
        });
    });
}
